### PR TITLE
CacheDictionaryUpdateQueue refactor

### LINKED ISSUE #29543: Deadlock in CacheDictionary
https://clickhouse-test-reports.s3.yandex.net/0/c37932a0b7af38ea40731a9421a4d9625254b2f1/stress_test_(thread).html#fail1

Interesting thread:
```
Thread 725 (Thread 0x7ff0e8f4c700 (LWP 19466)):
#0  0x00007ff435ab8376 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00000000096e15d0 in __tsan::call_pthread_cancel_with_cleanup(int (*)(void*), void (*)(void*), void*) ()
#2  0x000000000966d9e1 in pthread_cond_wait ()
#3  0x0000000019334a98 in Poco::EventImpl::waitImpl (this=0x7b2000076398) at ../contrib/poco/Foundation/src/Event_POSIX.cpp:106
#4  0x00000000097783c1 in Poco::Event::wait (this=0x7b2000076398) at ../contrib/poco/Foundation/include/Poco/Event.h:97
#5  ThreadFromGlobalPool::join (this=0x7b0c0144ce20) at ../src/Common/ThreadPool.h:217
#6  ThreadPoolImpl<ThreadFromGlobalPool>::finalize (this=this@entry=0x7b680161b598) at ../src/Common/ThreadPool.cpp:215
#7  0x00000000097780c5 in ThreadPoolImpl<ThreadFromGlobalPool>::~ThreadPoolImpl (this=0x7b680161b598) at ../src/Common/ThreadPool.cpp:201
#8  0x000000001207e696 in DB::CacheDictionaryUpdateQueue<(DB::DictionaryKeyType)0>::~CacheDictionaryUpdateQueue (this=0x7b680161b420) at ../src/Dictionaries/CacheDictionaryUpdateQueue.cpp:46
#9  0x00000000120675b2 in DB::CacheDictionary<(DB::DictionaryKeyType)0>::~CacheDictionary() ()
#10 0x000000001206761a in DB::CacheDictionary<(DB::DictionaryKeyType)0>::~CacheDictionary() ()
#11 0x000000001498eef5 in std::__1::default_delete<DB::IDictionary>::operator() (__ptr=0x7b680161b200, this=<optimized out>) at ../contrib/libcxx/include/memory:1397
#12 std::__1::__shared_ptr_pointer<DB::IDictionary*, std::__1::default_delete<DB::IDictionary>, std::__1::allocator<DB::IDictionary> >::__on_zero_shared (this=<optimized out>) at ../contrib/libcxx/include/memory:2565
#13 0x00000000149a0237 in std::__1::__shared_count::__release_shared (this=0x7b08001aa820) at ../contrib/libcxx/include/memory:2475
#14 std::__1::__shared_weak_count::__release_shared (this=0x7b08001aa820) at ../contrib/libcxx/include/memory:2517
#15 std::__1::shared_ptr<DB::IExternalLoadable const>::~shared_ptr (this=0x7b28014590a0) at ../contrib/libcxx/include/memory:3212
#16 DB::ExternalLoader::LoadingDispatcher::Info::~Info (this=0x7b2801459088) at ../src/Interpreters/ExternalLoader.cpp:707
#17 std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::ExternalLoader::LoadingDispatcher::Info>::~pair (this=0x7b2801459070) at ../contrib/libcxx/include/utility:297
#18 std::__1::destroy_at<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::ExternalLoader::LoadingDispatcher::Info> > (__loc=0x7b2801459070) at ../contrib/libcxx/include/__memory/base.h:118
#19 std::__1::allocator_traits<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*> > >::__destroy<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::ExternalLoader::LoadingDispatcher::Info> > (__p=__p@entry=0x7b2801459070) at ../contrib/libcxx/include/__memory/allocator_traits.h:547
#20 0x00000000149939eb in std::__1::allocator_traits<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*> > >::destroy<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::ExternalLoader::LoadingDispatcher::Info> > (__a=..., __p=0x80) at ../contrib/libcxx/include/__memory/allocator_traits.h:487
#21 std::__1::__hash_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*> > >::operator() (this=0x7ff0e8f024b8, __p=0x7b2801459060) at ../contrib/libcxx/include/__hash_table:849
#22 std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*>, std::__1::__hash_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*> > > >::reset (this=0x7ff0e8f024b0, __p=0x0) at ../contrib/libcxx/include/memory:1658
#23 std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*>, std::__1::__hash_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, void*> > > >::~unique_ptr (this=0x7ff0e8f024b0) at ../contrib/libcxx/include/memory:1612
#24 std::__1::__hash_table<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, std::__1::__unordered_map_hasher<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::__unordered_map_equal<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info> > >::erase (this=0x7b4800006ca8, __p=...) at ../contrib/libcxx/include/__hash_table:2498
#25 std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::ExternalLoader::LoadingDispatcher::Info, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::ExternalLoader::LoadingDispatcher::Info> > >::erase (this=0x7b4800006ca8, __p=...) at ../contrib/libcxx/include/unordered_map:1267
#26 DB::ExternalLoader::LoadingDispatcher::setConfiguration (this=this@entry=0x7b4800006c00, new_configs=...) at ../src/Interpreters/ExternalLoader.cpp:486
#27 0x00000000149900c0 in DB::ExternalLoader::reloadConfig (this=this@entry=0x7b9000000430, repository_name=...) at ../src/Interpreters/ExternalLoader.cpp:1498
#28 0x00000000149910d8 in DB::ExternalLoader::addConfigRepository(std::__1::unique_ptr<DB::IExternalLoaderConfigRepository, std::__1::default_delete<DB::IExternalLoaderConfigRepository> >) const::$_0::operator()() const (this=<optimized out>) at ../src/Interpreters/ExternalLoader.cpp:1291
#29 std::__1::__invoke<DB::ExternalLoader::addConfigRepository(std::__1::unique_ptr<DB::IExternalLoaderConfigRepository, std::__1::default_delete<DB::IExternalLoaderConfigRepository> >) const::$_0&> (__f=...) at ../contrib/libcxx/include/type_traits:3676
#30 std::__1::__invoke_void_return_wrapper<void>::__call<DB::ExternalLoader::addConfigRepository(std::__1::unique_ptr<DB::IExternalLoaderConfigRepository, std::__1::default_delete<DB::IExternalLoaderConfigRepository> >) const::$_0&>(DB::ExternalLoader::addConfigRepository(std::__1::unique_ptr<DB::IExternalLoaderConfigRepository, std::__1::default_delete<DB::IExternalLoaderConfigRepository> >) const::$_0&) (__args=...) at ../contrib/libcxx/include/__functional_base:348
#31 std::__1::__function::__default_alloc_func<DB::ExternalLoader::addConfigRepository(std::__1::unique_ptr<DB::IExternalLoaderConfigRepository, std::__1::default_delete<DB::IExternalLoaderConfigRepository> >) const::$_0, void ()>::operator()() (this=<optimized out>) at ../contrib/libcxx/include/functional:1608
#32 std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::ExternalLoader::addConfigRepository(std::__1::unique_ptr<DB::IExternalLoaderConfigRepository, std::__1::default_delete<DB::IExternalLoaderConfigRepository> >) const::$_0, void ()> >(std::__1::__function::__policy_storage const*) (__buf=<optimized out>) at ../contrib/libcxx/include/functional:2089
#33 0x00000000156c1b9f in std::__1::__function::__policy_func<void ()>::operator()() const (this=0x7b480155c250) at ../contrib/libcxx/include/functional:2221
#34 std::__1::function<void ()>::operator()() const (this=0x7b480155c250) at ../contrib/libcxx/include/functional:2560
#35 basic_scope_guard<std::__1::function<void ()> >::invoke() (this=0x7b480155c250) at ../base/common/../common/scope_guard.h:94
#36 basic_scope_guard<std::__1::function<void ()> >::reset() (this=0x7b480155c250) at ../base/common/../common/scope_guard.h:57
#37 DB::StorageDictionary::removeDictionaryConfigurationFromRepository (this=0x7b480155c100) at ../src/Storages/StorageDictionary.cpp:196
#38 DB::StorageDictionary::shutdown (this=0x7b480155c100) at ../src/Storages/StorageDictionary.cpp:177
#39 0x0000000014d97fe7 in DB::IStorage::flushAndShutdown (this=0x7b480155c100) at ../src/Storages/IStorage.h:485
#40 DB::InterpreterDropQuery::executeToTableImpl (this=<optimized out>, this@entry=0x7b0c0064dac0, query=..., db=..., uuid_to_wait=...) at ../src/Interpreters/InterpreterDropQuery.cpp:222
#41 0x0000000014d96c1d in DB::InterpreterDropQuery::executeToTable (this=0x80, this@entry=0x7b0c0064dac0, query=...) at ../src/Interpreters/InterpreterDropQuery.cpp:85
#42 0x0000000014d9669a in DB::InterpreterDropQuery::execute (this=0x7b0c0064dac0) at ../src/Interpreters/InterpreterDropQuery.cpp:62
#43 0x00000000153976f0 in DB::executeQueryImpl (begin=0x7b0c0064dac0 "X\266\357\004", begin@entry=0x7b0c00c92310 "DROP DICTIONARY dictdb_01076.dict_datarace;", end=end@entry=0x7b0c00c9233b "", context=..., internal=false, stage=DB::QueryProcessingStage::Complete, istr=0x0) at ../src/Interpreters/executeQuery.cpp:635
#44 0x00000000153954c2 in DB::executeQuery (query=..., context=..., internal=false, stage=<optimized out>) at ../src/Interpreters/executeQuery.cpp:950
#45 0x0000000015e3b287 in DB::TCPHandler::runImpl (this=this@entry=0x7b64006c0c00) at ../src/Server/TCPHandler.cpp:292
#46 0x0000000015e4a988 in DB::TCPHandler::run (this=0x7b64006c0c00) at ../src/Server/TCPHandler.cpp:1628
#47 0x0000000019250803 in Poco::Net::TCPServerConnection::start (this=0x7b64006c0c00) at ../contrib/poco/Net/src/TCPServerConnection.cpp:43
#48 0x0000000019251013 in Poco::Net::TCPServerDispatcher::run (this=<optimized out>) at ../contrib/poco/Net/src/TCPServerDispatcher.cpp:115
#49 0x00000000193d2d36 in Poco::PooledThread::run (this=<optimized out>) at ../contrib/poco/Foundation/src/ThreadPool.cpp:199
#50 0x00000000193d0f10 in Poco::(anonymous namespace)::RunnableHolder::run (this=<optimized out>) at ../contrib/poco/Foundation/src/Thread.cpp:55
#51 0x00000000193cf5e8 in Poco::ThreadImpl::runnableEntry (pThread=0x7b54001d99b8) at ../contrib/poco/Foundation/src/Thread_POSIX.cpp:345
#52 0x000000000966cccd in __tsan_thread_start_func ()
#53 0x00007ff435ab1609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#54 0x00007ff4359d8293 in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes #29543

### CODE REVIEW

File: src/Dictionaries/CacheDictionaryUpdateQueue.cpp [hunk 1]
```diff
@@ -35,40 +36,56 @@ CacheDictionaryUpdateQueue<dictionary_key_type>::CacheDictionaryUpdateQueue(
...
+        throw;
+    }
 }
 
 template <DictionaryKeyType dictionary_key_type>
 void CacheDictionaryUpdateQueue<dictionary_key_type>::waitForCurrentUpdateFinish(CacheDictionaryUpdateUnitPtr<dictionary_key_type> & update_unit_ptr) const
 {
     if (finished)
-        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue finished");
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue for dictionary {} already finished", dictionary_name_for_logs);
```
Comment (azat) [inline src/Dictionaries/CacheDictionaryUpdateQueue.cpp:84]: Why `UNSUPPORTED_METHOD`, looks like `LOGICAL_ERROR` is better?

File: src/Dictionaries/CacheDictionaryUpdateQueue.cpp [hunk 2]
```diff
@@ -108,13 +125,14 @@ void CacheDictionaryUpdateQueue<dictionary_key_type>::waitForCurrentUpdateFinish
 template <DictionaryKeyType dictionary_key_type>
 void CacheDictionaryUpdateQueue<dictionary_key_type>::stopAndWait()
 {
-    finished = true;
-    update_queue.clear();
+    if (finished.exchange(true))
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue for dictionary {} already finished", dictionary_name_for_logs);
```
Comment (azat) [inline src/Dictionaries/CacheDictionaryUpdateQueue.cpp:129]: Same here, `LOGICAL_ERROR`?

File: src/Dictionaries/CacheDictionaryUpdateQueue.h [hunk 3]
```diff
@@ -147,23 +152,29 @@ class CacheDictionaryUpdateQueue
     */
     void waitForCurrentUpdateFinish(CacheDictionaryUpdateUnitPtr<dictionary_key_type> & update_unit_ptr) const;
 
+    /// Get current update queue size
+    size_t getSize() const;
```
Comment (azat) [inline src/Dictionaries/CacheDictionaryUpdateQueue.h:156]: Looks like there is no need in exporting this for public use?

File: src/Dictionaries/CacheDictionaryUpdateQueue.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Dictionaries/CacheDictionaryUpdateQueue.cpp:84]: Fixed.

File: src/Dictionaries/CacheDictionaryUpdateQueue.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Dictionaries/CacheDictionaryUpdateQueue.cpp:129]: Fixed.

File: src/Dictionaries/CacheDictionaryUpdateQueue.h (hunk 3 above)
Comment (kitaisreal) [reply src/Dictionaries/CacheDictionaryUpdateQueue.h:156]: Fixed.

**Comment by kitaisreal** [comment]
Better fix in https://github.com/ClickHouse/ClickHouse/pull/29801.

**Pull request closed without merging**